### PR TITLE
0015 derive seq

### PR DESCRIPTION
**Pull request by ramiromagno**
Thank you for your Pull Request! We have developed this task checklist from the
[Development Process
Guide](https://pharmaverse.github.io/sdtm.oak/CONTRIBUTING.html)
to help with the final steps of the process. Completing the below tasks helps to
ensure our reviewers can maximize their time on your code as well as making sure
the oak codebase remains robust and consistent.

Please check off each taskbox as an acknowledgment that you completed the task
or check off that it is not relevant to your Pull Request. This checklist is
part of the Github Action workflows and the Pull Request will not be merged into
the `devel` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull z
Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the 
[tidyverse style guide](https://style.tidyverse.org/). Run 
`styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should
consider realistic data scenarios and edge cases, e.g. empty datasets, errors,
boundary cases etc. - See 
[Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you 
fully follow the 
[deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords
and families. Refer to the 
[categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the
`NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` if the changes pertain to a user-facing function (i.e. it
has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build oak site `pkgdown::build_site()` and check that all affected
examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

File: R/derive_seq.R [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+#'
+#' Prior to the derivation of `tgt_var`, the data frame `tgt_dat` is sorted
+#' according to grouping variables indicated in `rec_vars`.
+#'
+#' @param tgt_dat The target dataset, a data frame.
+#' @param tgt_var The target SDTM variable: a single string indicating the name
+#'   of the sequence number (`--SEQ`) variable, e.g. `"DSSEQ"`. Note that
+#'   supplying a name not ending in `"SEQ"` will raise a warning.
+#' @param rec_vars A character vector of record-level identifier variables. Use
+#' [rec_vars()] to quickly get the identifier variables for a given domain.
```
Comment (rammprasad) [inline R/derive_seq.R:19]: Does rec_vars() pull the default values from "data-raw/domain_record_vars.csv.xz"? Does this file come from {roak}? If so, I will make it a CSV and make it viewable to the users. There is no guarantee that the keys mentioned in the file are accurate.

Comment (rammprasad) [reply R/derive_seq.R:19]: Also, I think we need to document this as sample key variables.

File: data-raw/pharmaversesdtm.R [hunk 2]
```diff
@@ -0,0 +1,36 @@
+# Title: SDTM domain datasets provided by R package pharmaversesdtm
+# Description: This script is used to snatch those SDTM datasets provided by
+#              pharmaversesdtm. These include datasets that are therapeutic area
+#              (TA)-agnostic (DM, VS, EG, etc.) as well TA-specific ones
+#              (RS, TR, OE, etc.). We export them to inst/domain as RDS.
+#
```
Comment (rammprasad) [inline data-raw/pharmaversesdtm.R:6]: Is there a reason to copy these datasets to {sdtm.oak}? We can probably add `pharmaversesdtm` as a dependency and load it when needed. I highly doubt we will need all these datasets.

File: data-raw/roak_ex.R [hunk 3]
```diff
@@ -0,0 +1,31 @@
+# Title: `roak` example data sets
+# Description: Data sets provided during the initial development of sdtm.oak
+# by the team at Roche.
```
Comment (rammprasad) [inline data-raw/roak_ex.R:3]: Can we remove {roak} references in {sdtm.oak}. Users may not understand the purpose of this.

**Review by rammprasad — CHANGES_REQUESTED** [review]
Overall function works as expected. We just need to look into the data-raw folder and keep just the ones we need.

File: R/derive_seq.R (hunk 1 above)
Comment (ramiromagno) [reply R/derive_seq.R:19]: - `rec_vars()` pulls the default values from an internal data set that becames bundled with the package. The source for this data set is explained here: https://github.com/pharmaverse/sdtm.oak/pull/53/commits/8662fa713175a3eeeeb6dcfc64e82093e5213dc8#diff-d23bac3cea2c5fe24e84e8c626ec809f57783e96bfa6e7b606eda309dfadc994R1-R4.
- Instead of making the CSV available to users, an exported data set would be more appropriate. I can change that internal data set to an exported data set.
- If those default keys are not appropriate for the user specific case, they can always pass it as there is an argument `rec_vars()`.
- Regarding the naming of those key variables as "samples" key variables, I tried to follow CDISC SDTM terminology, and to my understanding those unique rows to be identified as referred to as records in CDISC terminology. So I will leave it be and let you confirm.

File: data-raw/roak_ex.R (hunk 3 above)
Comment (ramiromagno) [reply data-raw/roak_ex.R:3]: All source files in `data-raw/` directory are not available to users. So they won't be confused unless they read the source code of the package. Those three lines you mention are only for developers' own reference.

File: data-raw/pharmaversesdtm.R (hunk 2 above)
Comment (ramiromagno) [reply data-raw/pharmaversesdtm.R:6]: The only reason was to have ownership over these data. As these data become used for testing and examples, an upstream change might suddenly break the building of `{sdtm.oak}`.

**Review by rammprasad — APPROVED** [review]
Thank you for addressing the comments.

**Comment by rammprasad** [comment]
@edgar-manukyan - Ready for your review.

**Review by edgar-manukyan — APPROVED** [review]
Thanks so much Ramiro, looks fine to me. Just left two comments in case you want to address them. Are we going to add tests at some point?

File: R/derive_seq.R [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    admiraldev::assert_character_vector(sbj_vars)
+    admiraldev::assert_data_frame(tgt_dat,
+      required_vars = rlang::syms(rec_vars),
+      optional = FALSE
+    )
+
+    admiraldev::assert_integer_scalar(start_at, subset = "non-negative")
+
+    tgt_dat |>
+      dplyr::ungroup() |> # ensure that is ungrouped
```
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/derive_seq.R:63]: The command following the ungroup is a `dplyr::arrange()`, so I am just making sure that the arrange is not happening on a group by group basis if that was the case. So just trying to be defensive here.

File: R/derive_seq.R [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+#' This function adds a newly derived variable to `tgt_dat`, namely the sequence
+#' number (`--SEQ`) whose name is the one provided in `tgt_var`. An integer
+#' sequence is generated that uniquely identifies each record within the domain.
+#'
+#' Prior to the derivation of `tgt_var`, the data frame `tgt_dat` is sorted
+#' according to grouping variables indicated in `rec_vars`.
+#'
+#' @param tgt_dat The target dataset, a data frame.
+#' @param tgt_var The target SDTM variable: a single string indicating the name
+#'   of the sequence number (`--SEQ`) variable, e.g. `"DSSEQ"`. Note that
```
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/derive_seq.R:16]: @edgar-manukyan: Was this comment for me, or perhaps for @rammprasad ?

**Comment by ramiromagno** [comment]
> Thanks so much Ramiro, looks fine to me. Just left two comments in case you want to address them. Are we going to add tests at some point?

I am happy to write the unit tests, but could you provide some input and output examples such that the tests are based on realistic examples, and also cover corner cases. (you guys have more experience here).

File: R/derive_seq.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (edgar-manukyan) [reply R/derive_seq.R:16]: I was just reading the function header and it says:
>  Note that  supplying a name not ending in `"SEQ"` will raise a warning.

The I supplied a name that did not end with SEQ expecting a warning, but there was no warning.

File: R/derive_seq.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (edgar-manukyan) [reply R/derive_seq.R:63]: Then perhaps the comment can say "ensure that no prior grouping exists that alters ordering and new grouping".

File: R/derive_seq.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/derive_seq.R:16]: Okay, so you agree that a warning is appropriate, correct?

[the remark this answers is not in the document]
Comment (edgar-manukyan) [reply R/derive_seq.R:16]: I think that would make sense and if we are documenting it then we better do it 👍

[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/derive_seq.R:16]: Yeah, I had initially planned to do it, so I documented it but then I did not implement it because I felt it would be better to ask you guys... And then I forgot and the docs remained...
